File: src/js/src/sniper.js
const t = require("../util/tester.js")("Sniper Test"), fs = require("fs"), u = require("util");

function reader(path) { return fs.readFileSync(path, "utf-8");}
function writer(path, msg){ fs.writeFileSync(path, msg, "utf-8");}

class Sniper {
  constructor(items){
    let [nm, it,sp,lp,lb,st] = items.split(' '); this.name = nm; this.item = it; this.state = st.trim();
    this.stopPrice = parseInt(sp); this.lastPrice = parseInt(lp); this.lastBid = parseInt(lb);
    this.msg = "toAuctionMessage is None";
  }
  close(){ this.state = this.state == "WINNING" ? "WON": "LOST";}
  price(price, increment, bidder){
    if(this.name === bidder) { this.lastPrice = price; this.state = "WINNING"; return;}
    let bid = price + increment; if(bid > this.stopPrice){ this.lastPrice = price; this.state = "LOSING"; return;}
    this.lastPrice = price; this.lastBid = bid; this.state = "BIDDING";
    this.msg = u.format("SOLVersion: 1.1; Command: BID; Price: %d;", bid);
  }
  fail(msg){ this.msg = "failed: " + msg;  this.lastPrice = 0; this.lastBid = 0; this.state = "FAILED";}
  res(){ return `${this.msg}\n${this.item} ${this.lastPrice} ${this.lastBid} ${this.state}`;}
} exports.Sniper = Sniper;

class Translator {
  constructor(msg, sniper){ this.msg = msg; this.sniper = sniper; this.events = {};}
  mkEvent(){
    let events = this.msg.split(';'); events.pop();
    events.forEach(pair => { let [k,v] = pair.split(':'); this.events[k.trim()] = v.trim();});
  }
  get(key){
    let val = this.events[key]; if (val == null || val === ""){ throw new Error("missing field: " + key);}
    return val;
  }
  event(){ return this.get("Event")} getInt(key){ return parseInt(this.get(key));} bidder() { return this.get("Bidder");}
  price(){ return this.getInt("CurrentPrice");} increment(){ return this.getInt("Increment");}
  translate(){
    try {
      this.mkEvent(); let type = this.event();
      if(type === "CLOSE"){ this.sniper.close(); return;}
      if(type === "PRICE"){ this.sniper.price( this.price(), this.increment(), this.bidder()); return;}
    }catch(e){ this.sniper.fail(`msg is [${this.msg}] ${e}`);}
  }
} exports.Translator = Translator;

function domain(items, msg){
  let sniper = new Sniper(items), trans = new Translator(msg, sniper);
  trans.translate(); return sniper.res();
}

function test(){
  let res = domain("sniper item-123 1300 1098 1098 WINNING", "SOLVersion: 1.1; Event: CLOSE;");
  console.log(res);
}

function refactor(){
  let [items, msg ] = reader("a:/pj/sniper/rr/req1.txt").split('\n');
  let res = domain(items, msg), exp = reader("a:/pj/sniper/rr/res2.txt");
  res = res.replace("\n", " "); exp = exp.replace("\r\n", " ");
  t.test(res.trim(), exp.trim(), "今回リファクターではあえて間違いテストをチェックしたい");
}

function developer(arg, exp, disp){
  let [items, msg] = arg.split('\n'), ans = domain(items, msg);
  t.test(ans.replace("\n", " "), exp.replace("\r\n", " "), disp);
}
function develop(){
  const arg1 = reader("a:/pj/sniper/rr/req1.txt"), exp1 = reader("a:/pj/sniper/rr/res1.txt");
  const arg2 = reader("a:/pj/sniper/rr/req2.txt"), exp2 = reader("a:/pj/sniper/rr/res2.txt");
  const arg3 = reader("a:/pj/sniper/rr/req3.txt"), exp3 = reader("a:/pj/sniper/rr/res3.txt");
  const arg4 = reader("a:/pj/sniper/rr/req4.txt"), exp4 = reader("a:/pj/sniper/rr/res4.txt");
  const arg5 = reader("a:/pj/sniper/rr/req5.txt"), exp5 = reader("a:/pj/sniper/rr/res5.txt");
  const arg6 = reader("a:/pj/sniper/rr/req6.txt"), exp6 = reader("a:/pj/sniper/rr/res8.txt");
  developer(arg1,exp1, "参加要請中に限度額以下の価格情報がくると入札中になって買い注文を出すよ");
  developer(arg2,exp2, "参加要請中にオークションが閉まると落札は失敗で買い注文は無しだ");
  developer(arg3,exp3, "入札中にオークションが閉まっても落札は失敗で買い注文は無しだ");
  developer(arg4,exp4, "一位入札中にオークションが閉まれば落札は成功で買い注文は無しだ");
  developer(arg5,exp5, "一位入札中でも限度額以上の価格情報がくれば脱落中で買い注文は無しだ");
  developer(arg6,exp6, "ヘンテコなメッセージが来ると金額はすべて０にしてエラー発生だ");
}

function product(){
  let [items, msg] = reader("toSniper.txt").split('\n');
  writer("fromSniper.txt", domain(items, msg));
}

if(require.main === module){
  // test()
  // refactor()
  // develop()
  product()
}

/*
node src/js/src/sniper.js
 */

File: src/haskell/src/Sniper.hs
module Src.Haskell.Src.Sniper(sniper, translator, job)where
import Data.List
import qualified Data.Map as M

import Src.Haskell.Util.Tester(zz_act)
import Src.Haskell.Util.Util(trim, split)

bid_msg = "SOLVersion: 1.1; Command: BID; Price: "

data Sniper = Sniper {
  name :: String, item :: String, stopPrice :: Int, lastPrice :: Int, lastBid :: Int, state :: String, aucMsg :: String
}

instance Show Sniper where
  show s = (aucMsg s) ++ "\n" ++ (item s) ++ " " ++ (show(lastPrice s)) ++ " " ++ (show(lastBid s)) ++ " " ++ (state s)

sniper:: [String] -> Sniper; sniper(n:i:s:p:b:t:_) = Sniper n i (read s) (read p) (read b) t "toAuctionMessage is None"

mapping :: String -> M.Map String String; mapping msg = M.fromList(map(pair) (init(split msg ";"))) where
  pair f = tpls(split f ":"); tpls (k:v:_) = if k=="" || v=="" then("","") else (trim k, trim v)

checker :: Maybe String -> String; checker ms = maybe "Failed" id ms

data Translator = Close | Price Int Int String | Failed String deriving Show

missingField :: String -> String -> Translator
missingField msg field = aboutFailed $"msg is [" ++ msg ++ "] missing field [" ++ field ++ "]"

aboutPrice :: M.Map String String -> String -> Translator
aboutPrice maps msg = 
  let price =  checker (M.lookup "CurrentPrice" maps) in if price ==  "Failed"  then missingField msg "CurrentPrice" else
  let inc =    checker (M.lookup "Increment" maps)    in if inc ==    "Failed"  then missingField msg "Increment"    else
  let bidder = checker (M.lookup "Bidder" maps)       in if bidder == "Failed"  then missingField msg "Bidder"       else
  Price (read price) (read inc) bidder

aboutFailed :: String -> Translator; aboutFailed msg = Failed ("failed: " ++ msg)

translator :: String -> Translator
translator msg = let maps = mapping msg in case checker (M.lookup "Event" maps) of
  "CLOSE" -> Close; "PRICE" -> aboutPrice maps msg; other   -> missingField msg "Event"

close :: Sniper -> Sniper; close s | (state s) == "WINNING" = s {state = "WON"} | otherwise = s {state = "LOST"}

price :: Sniper -> Int -> Int -> String -> Sniper
price s p i b = if b == (name s) then s {lastPrice = p, state = "WINNING"} else
  let bid = p + i in if bid > (stopPrice s) then s {lastPrice = p, state = "LOSING"} else
    s { lastPrice = p, lastBid = bid, state = "BIDDING", aucMsg = bid_msg ++ (show bid) ++ ";"}

failed :: Sniper -> String -> Sniper; failed s m = s {lastPrice = 0, lastBid = 0, state = "FAILED", aucMsg = m}

job :: Sniper -> Translator -> String
job s t = case t of Close -> show (close s); (Price p i b) -> show $ price s p i b; (Failed m) -> show $ failed s m

domain :: [String] -> String; domain (items:msg:_) =
  let s = sniper (words items)
      t = translator msg
  in job s t

test = let bidding = "sniper item-123 3000 1000 1098 BIDDING"
           price_other = "SOLVersion: 1.1; Event: PRICE; CurrentPrice: 1000; Increment: 98; Bidder: someone else;"
       in domain [bidding, price_other]

retToSp :: String -> String; retToSp s = let (a:b:_) = lines s in a ++ " " ++ b
refactor = do
  args <- readFile "a:/pj/sniper/rr/req1.txt"; exp <- readFile "a:/pj/sniper/rr/res2.txt"
  let act = domain(lines args)
  return (zz_act "Sniper Test" (retToSp act) (retToSp exp) "it makes snse to fail by Refactor")

developer args exp name = do
  let act = domain(lines args)
  return (zz_act "Sniper" (retToSp act) (retToSp exp) name)
develop = do
  req1 <- readFile "a:/pj/sniper/rr/req1.txt"; res1 <- readFile "a:/pj/sniper/rr/res1.txt"
  req2 <- readFile "a:/pj/sniper/rr/req2.txt"; res2 <- readFile "a:/pj/sniper/rr/res2.txt"
  req3 <- readFile "a:/pj/sniper/rr/req3.txt"; res3 <- readFile "a:/pj/sniper/rr/res3.txt"
  req4 <- readFile "a:/pj/sniper/rr/req4.txt"; res4 <- readFile "a:/pj/sniper/rr/res4.txt"
  req5 <- readFile "a:/pj/sniper/rr/req5.txt"; res5 <- readFile "a:/pj/sniper/rr/res5.txt"
  req6 <- readFile "a:/pj/sniper/rr/req6.txt"; res6 <- readFile "a:/pj/sniper/rr/res9.txt"
  ans1 <- developer req1 res1 "WnenReceivedBuiableInfoThenBuyAndJoining"
  ans2 <- developer req2 res2 "WnenJoiningIfReceivedCloseInfoWhileNoMsgAndLost"
  ans3 <- developer req3 res3 "WnenBiddingIfReceivedCloseInfoWhileNoMsgAndLost"
  ans4 <- developer req4 res4 "WnenWinningIfReceivedCloseInfoWhileNoMsgAndWon"
  ans5 <- developer req5 res5 "WnenWinningIfReceivedThooHiPriceInfoWhileNoMsgAndLosing"
  ans6 <- developer req6 res6 "WnenWinningIfReceivedFunnyMsgWhileFailMsgAndFailed"
  return ( ans1 ++ "\n" ++ ans2 ++ "\n" ++ ans3 ++ "\n" ++ ans4 ++ "\n" ++ ans5 ++ "\n" ++ ans6)

release :: IO (); release = do
  arg <- readFile "toSniper.txt"
  let ans = domain (lines arg)
  writeFile "fromSniper.txt" ans
  return ()

main = do
  -- let ans = test
  -- ans <- refactor
  -- ans <- develop
  release
  -- putStrLn ans

{-
runghc src/haskell/src/Sniper.hs
:l src/haskell/src/Sniper.hs
cd src/haskell
cd ../..
runghc src/Sniper.hs
-}